faunadb-android/src/main/java/com/faunadb/client/FaunaClient.java

```diff
@@ -0,0 +1,288 @@
+    private OkHttpClient httpClient;
+
+    private Builder() {
+    }
+
+    /**
+     * Sets the secret to be passed to FaunaDB as a authentication token
+     *
+     * @param secret the auth token secret
+     * @return this {@link Builder} object
+     */
+    public Builder withSecret(String secret) {
+      this.secret = secret;
+      return this;
+    }
+
+    /**
+     * Sets the FaunaDB endpoint url for the built {@link FaunaClient}.
+     *
+     * @param endpoint the root endpoint URL
+     * @return this {@link Builder} object
+     * @throws java.net.MalformedURLException when an invalid string is passed
+     */
+    public Builder withEndpoint(String endpoint) throws MalformedURLException {
+      this.endpoint = new URL(endpoint);
+      return this;
+    }
+
+    /**
+     * Sets a custom {@link OkHttpClient} implementation that the built {@link FaunaClient} will use.
+     * This custom implementation can be provided to control the behavior of the underlying HTTP transport.
+     *
+     * @param httpClient the custom {@link OkHttpClient} instance
+     * @return this {@link Builder} object
+     */
+    public Builder withHttpClient(OkHttpClient httpClient) {
+      this.httpClient = httpClient;
+      return this;
+    }
+
+    /**
+     * Returns a newly constructed {@link FaunaClient} with configuration based on the settings of this {@link Builder}.
+     *
+     * @return a new {@link FaunaClient}
+     */
+    public FaunaClient build() {
+      if (httpClient == null) {
+        return new FaunaClient(new OkHttpClient.Builder()
+                .connectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                .readTimeout(DEFAULT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                .writeTimeout(DEFAULT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                .retryOnConnectionFailure(true)
+                .build(), endpoint, secret);
+      }
+
+      return new FaunaClient(httpClient, endpoint, secret);
+    }
+  }
+
+  private final ObjectMapper json = new ObjectMapper().registerModule(new GuavaModule());
+  private final OkHttpClient client;
+  private final URL endpoint;
+  private final String authHeader;
+
+  private FaunaClient(OkHttpClient client, URL endpoint, String secret) {
+    this.client = client;
+    this.endpoint = endpoint;
+    this.authHeader = Credentials.basic(secret, "");
+  }
+
+  /**
+   * Creates a session client with the user secret provided. Queries submited to a session client will be
+   * authenticated with the secret provided.
+   *
+   * @param secret user secret for the session client
+   * @return a new {@link FaunaClient}
+   */
+  public FaunaClient newSessionClient(String secret) {
+    return new FaunaClient(client, endpoint, secret);
+  }
+
+  /**
+   * Frees any resources held by the client. Also closes the underlying {@link OkHttpClient}.
+   */
+  public void close() throws IOException {
+    client.dispatcher().executorService().shutdown();
```

Comment:
You need a ref count. You can only close the http resources once all clients are closed.
Otherwise there is a change a session client gets closed and the parent client becomes invalid.

Also, we should make the class implement `AutoCloseable`.